locksmith/src/controllers/v2/hooksController.ts

```diff
@@ -60,3 +64,97 @@ export const guildHook: RequestHandler = async (request, response) => {
+      return a
+    }, z.array(z.string()))
+    // transform each recipient address using a normalization function
+    .transform((items) =>
+      items.map((item) => normalizer.ethereumAddress(item))
+    ),
+})
+
+/**
+ * The `gitcoinHook` function is designed to handle the scoring and verification process for recipients (typically wallet addresses)
+ * submitted through a request. This process involves several key steps:
+ * 1. Parsing and validating the input parameters from the request, including network details, recipient addresses, and lock address.
+ * 2. Submitting each recipient's address for scoring, initiating an asynchronous process that calculates their respective scores
+ *    based on predefined criteria (via `submitAddressForScoring`).
+ * 3. Retrieving a wallet that is capable of signing messages, which is necessary for generating cryptographic signatures as a form of
+ *    verification for recipients with scores above a certain threshold.
+ * 4. Fetching the scores for all submitted recipients, utilizing a batch or collective approach for efficiency.
+ * 5. Evaluating each recipient's score to determine if it exceeds the defined threshold (in this case, a score greater than 20).
+ *    Recipients with qualifying scores are then provided with a signature generated using the retrieved wallet, serving as a form
+ *    of validation or access granting. Recipients not meeting the score requirement receive an empty string, indicating a lack of
+ *    qualification.
+ * 6. Compiling and responding with the results, which include either signatures for qualified recipients or empty strings for
+ *    those who did not qualify, effectively communicating the outcome of the verification process.
+ *
+
+ */
+export const gitcoinHook: RequestHandler = async (request, response) => {
+  const { network, recipients, lockAddress } =
+    await gitcoinHookQuery.parseAsync(request.query)
+
+  try {
+    // submit each recipient for scoring
+    await Promise.all(
+      recipients.map((recipient) => submitAddressForScoring(recipient))
+    )
+
+    // retrieve the wallet for signing, ensuring it's available before generating signatures
+    const wallet = await getSignerFromOnKeyPurchaserHookOnLock({
+      lockAddress,
+      network,
+    })
+
+    if (!wallet) {
+      return response.status(422).json({
+        error: 'This lock has a misconfigured Gitcoin Passport hook.',
+      })
+    }
+
+    // here, we wait for 3 seconds after submission, as the scores are being processed
+    await new Promise((resolve) => setTimeout(resolve, 3000))
+
+    // retrieve scores for all submitted recipients
+    const scoresResponse = await checkMultipleScores(recipients)
+
+    // generate signatures for recipients with valid scores
+    const generatedSignatures = scoresResponse.map((recipient: any) => {
+      // only sign recipients who have a score greater than or equal to 20
+      if (recipient && recipient.score >= 20) {
```

Comment:
Should we make that score customizable? The lock manager when they set the hook would define what minimum score they want to allow. We already have `lockSettings` which can be used to store that (they store the passwords/discount codes)

Reply:
was thinking of just that. yes we can. will get right to implementing that

Reply:
done.

here's what the new UI looks like to enable the gitcoin passport hook

![improved UI](https://github.com/unlock-protocol/unlock/assets/46839250/8fe99a9f-5c6b-4316-97c9-99c53a7a0423)